metagraph/src/graph/alignment/aligner_chainer.cpp

```diff
@@ -0,0 +1,474 @@
+    if (begin == end || chain_end == query.data() + query.size()) {
+        callback(std::move(chain));
+        return;
+    }
+
+    score_t score = chain.get_score();
+
+    bool called = false;
+    for (auto it = begin; it != end; ++it) {
+        if (it->get_offset())
+            continue;
+
+        const char *next_begin = it->get_query().data();
+
+        assert(chain_begin - chain.get_clipping() == next_begin - it->get_clipping());
+        assert(it->get_orientation() == chain.get_orientation());
+
+        const char *next_end = next_begin + it->get_query().size();
+
+        if (next_begin <= chain_begin || next_end == chain_end)
+            continue;
+
+        if (chain.label_columns.size()
+                && !utils::share_element(it->label_columns.begin(),
+                                         it->label_columns.end(),
+                                         chain.label_columns.begin(),
+                                         chain.label_columns.end())) {
+            continue;
+        }
+
+        Alignment aln(*it);
+
+        if (next_begin >= chain_end) {
+            // no overlap
+            aln.insert_gap_prefix(next_begin - chain_end, node_overlap, config);
+
+        } else {
+            // trim, then fill in dummy nodes
+            assert(chain.get_end_clipping());
+
+            // first trim front of the incoming alignment
+            size_t overlap = std::min(
+                static_cast<size_t>((chain.get_cigar().data().end() - 2)->second),
+                aln.trim_query_prefix(chain_end - it->get_query().data(), node_overlap, config)
+            );
+
+            if (aln.empty() || aln.get_sequence().size() <= node_overlap
+                    || (aln.get_cigar().data().begin()
+                            + static_cast<bool>(aln.get_clipping()))->first != Cigar::MATCH) {
+                continue;
+            }
+
+            assert(aln.get_query().data() == chain.get_query().data() + chain.get_query().size());
+
+            if (overlap < node_overlap)
+                aln.insert_gap_prefix(-overlap, node_overlap, config);
+        }
+
+        if (aln.empty())
+            continue;
+
+        score_t next_score = score + aln.get_score();
+        if (next_score > best_score[next_end - query.data()]) {
+            best_score[next_end - query.data()] = next_score;
+
+            // use append instead of splice because any clipping in aln represents
+            // internally clipped characters
+            Alignment next_chain(chain);
+            next_chain.trim_end_clipping();
+            bool modified = next_chain.append(std::move(aln));
+            assert(next_chain.get_score() == next_score);
+            if (next_chain.size()) {
+                called |= modified;
+                construct_alignment_chain<AlignmentCompare>(
+                    node_overlap, config, query, std::move(next_chain),
+                    it + 1, end, best_score, callback
+                );
```

Comment:
This can be rewritten in a way that doesn't involve recursion, but it's a bit more complicated. The `best_score` vector ensures that a lot of the calls get cut off early, but I suppose it's possible for the depth to be exceeded in a very bad case.